### PR TITLE
felix/bpf: include RPF and wg chains again.

### PR DESCRIPTION
**Pull request by tomastigera**
Since commit
https://github.com/projectcalico/calico/commit/cabb7c129bd476e9fc339bd02d285e457ed2993f
the prerouting raw chain does not include the RPF and wg chains as those
are removed by the chains generated by StaticBPFModeRawChains(). Hence
we do not enforce RPF when RPF is disabled or set to lose.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: felix/dataplane/linux/int_dataplane.go [hunk 1]
```diff
@@ -1265,6 +1266,11 @@ func (d *InternalDataplane) setUpIptablesBPF() {
 	}
 
 	for _, t := range d.iptablesRawTables {
+		if t.IPVersion != 4 {
+			continue
+		}
```
Comment (neiljerram) [inline felix/dataplane/linux/int_dataplane.go:1271]: Is this intended to stay?

File: felix/rules/static.go [hunk 2]
```diff
@@ -986,36 +986,8 @@ func (r *DefaultRuleRenderer) StaticRawTableChains(ipVersion uint8) []*Chain {
...
-			// yet.
-			Rule{
-				Action: SetMarkAction{Mark: 0},
-			},
-			// Now ensure that the packet is not tracked.
-			Rule{
-				Action: NoTrackAction{},
-			},
-		},
-	}
```
Comment (neiljerram) [inline felix/rules/static.go:1016]: As discussed, I think it would be nicer to keep this code here, and instead move the RPF and Wireguard rules here too, and call this for both IPv4 and IPv6.

**Review by neiljerram — CHANGES_REQUESTED** [review]

**Comment by tomastigera** [comment]
@neiljerram I force-pushed as it is a complete rework. Also includes a test that works for any dp.

File: felix/fv/rpf_test.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			pol.Spec.Ingress = []api.Rule{{Action: "Allow"}}
+			pol.Spec.Egress = []api.Rule{{Action: "Allow"}}
+			pol.Spec.Selector = "all()"
+
+			pol = func(policy *api.GlobalNetworkPolicy) *api.GlobalNetworkPolicy {
+				log.WithField("policy", dumpResource(policy)).Info("Creating policy")
+				policy, err := calicoClient.GlobalNetworkPolicies().Create(utils.Ctx, policy, utils.NoOptions)
+				Expect(err).NotTo(HaveOccurred())
+				return policy
+			}(pol)
```
Comment (neiljerram) [inline felix/fv/rpf_test.go:79]: That's a curious way of writing things - is there a clever reason?

File: felix/fv/rpf_test.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+				err := eth20.Start()
+				Expect(err).NotTo(HaveOccurred())
+
+				// assign address to eth20 and add route to the .20 network
+				felixes[0].Exec("ip", "route", "add", "192.168.20.0/24", "dev", "eth20")
+				felixes[0].Exec("ip", "addr", "add", "10.0.0.20/32", "dev", "eth20")
+				_, err = eth20.RunCmd("ip", "route", "add", "10.0.0.20/32", "dev", "eth0")
+				Expect(err).NotTo(HaveOccurred())
+				// Add a route to the test workload to the fake external
+				// client emulated by the test-workload
```
Comment (neiljerram) [inline felix/fv/rpf_test.go:132]: Can you reword this to distinguish more clearly between "test workload" (`w` I think) and "test-workload" (`eth20` I think).

File: felix/fv/rpf_test.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+				matcher := fmt.Sprintf("IP %s\\.30446 > %s\\.30446: UDP", fakeWorkloadIP, w.IP)
+				tcpdump.AddMatcher("UDP-30446", regexp.MustCompile(matcher))
+				tcpdump.Start()
+				defer tcpdump.Stop()
+
+				_, err := eth20.RunCmd("/pktgen", fakeWorkloadIP, w.IP, "udp",
+					"--port-src", "30446", "--port-dst", "30446", "--ip-id", "666")
+				Expect(err).NotTo(HaveOccurred())
+
+				// Expect to receive the packet from the .20 as the routing is correct
```
Comment (neiljerram) [inline felix/fv/rpf_test.go:156]: Comment doesn't seem to match what the check is doing.  (I.e. that there isn't anything in tcpdump)

File: felix/fv/rpf_test.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+				tcpdump := w.AttachTCPDump()
+				tcpdump.SetLogEnabled(true)
+				matcher := fmt.Sprintf("IP %s\\.30446 > %s\\.30446: UDP", fakeWorkloadIP, w.IP)
+				tcpdump.AddMatcher("UDP-30446", regexp.MustCompile(matcher))
+				tcpdump.Start()
+				defer tcpdump.Stop()
+
+				_, err := eth20.RunCmd("/pktgen", fakeWorkloadIP, w.IP, "udp",
+					"--port-src", "30446", "--port-dst", "30446", "--ip-id", "666")
```
Comment (neiljerram) [inline felix/fv/rpf_test.go:153]: What is the IP ID for here?

**Review by neiljerram — CHANGES_REQUESTED** [review]
Basically good, but worried about IPv6 and a few queries.

File: felix/fv/rpf_test.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/fv/rpf_test.go:1]: oops maybe I did not install them

File: felix/fv/rpf_test.go (hunk 6 above)
Comment (tomastigera) [reply felix/fv/rpf_test.go:153]: purely to be able to track the packet when debugging across interfaces and namespaces. I left it here in case there are future issues.

File: felix/dataplane/linux/int_dataplane.go [hunk 8]
```diff
@@ -1265,59 +1265,9 @@ func (d *InternalDataplane) setUpIptablesBPF() {
...
-			Name:  rules.ChainRawPrerouting,
-			Rules: rawRules,
-		}}
-		t.UpdateChains(rawChains)
-
-		t.InsertOrAppendRules("PREROUTING", []iptables.Rule{{
-			Action: iptables.JumpAction{Target: rules.ChainRawPrerouting},
-		}})
-
 		if t.IPVersion == 4 {
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/dataplane/linux/int_dataplane.go:1268]: I wasn't sure whether it matters. We support IPv4 only, but I can make it apply to v6 too.

File: felix/fv/rpf_test.go (hunk 3 above)
Comment (tomastigera) [reply felix/fv/rpf_test.go:79]: nope, me just taking a copy-pasta shortcut, will  redo it.

File: felix/dataplane/linux/int_dataplane.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/dataplane/linux/int_dataplane.go:1272]: > IP version 4 check is only because there was previously a version 4 check for StaticRawOutputChains.

Only this should be IPv4 now again.

File: felix/fv/rpf_test.go [hunk 9]
```diff
@@ -38,6 +39,13 @@ var _ = infrastructure.DatastoreDescribe(
 	"RPF tests",
 	[]apiconfig.DatastoreType{apiconfig.Kubernetes},
 	func(getInfra infrastructure.InfraFactory) {
+
+		// XXX disable for non-bpf as it may not be enforced
```
Comment (neiljerram) [inline felix/fv/rpf_test.go:43]: nit: what does "XXX" mean?

File: felix/fv/rpf_test.go (hunk 4 above)
Comment (neiljerram) [reply felix/fv/rpf_test.go:137]: Sorry, I wasn't clear, but I was hoping for a clarification of the comment wording...

File: felix/fv/rpf_test.go (hunk 6 above)
Comment (neiljerram) [reply felix/fv/rpf_test.go:165]: I'm still unclear though - does the following tcpdump match check for ID 666?  Or is the ID 666 only for manual debugging work?

**Review by neiljerram — CHANGES_REQUESTED** [review]
I'm afraid I'm still not clear on the ID 666 detail...

File: felix/dataplane/linux/int_dataplane.go [hunk 10]
```diff
@@ -1219,62 +1220,12 @@ func (d *InternalDataplane) setUpIptablesBPF() {
...
-			Action: iptables.JumpAction{Target: rpfChain[0].Name},
-		})
-
-		rawChains := []*iptables.Chain{{
-			Name:  rules.ChainRawPrerouting,
-			Rules: rawRules,
-		}}
-		t.UpdateChains(rawChains)
-
+		t.UpdateChains(d.ruleRenderer.StaticBPFModeRawChains(
```
Comment (fasaxc) [inline felix/dataplane/linux/int_dataplane.go:1223]: Thanks for the cleanup!

File: felix/fv/rpf_test.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+	. "github.com/projectcalico/calico/felix/fv/connectivity"
+	"github.com/projectcalico/calico/felix/fv/infrastructure"
+	"github.com/projectcalico/calico/felix/fv/utils"
+	"github.com/projectcalico/calico/felix/fv/workload"
+	"github.com/projectcalico/calico/libcalico-go/lib/apiconfig"
+	client "github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+)
+
+var _ = infrastructure.DatastoreDescribe(
+	"RPF tests",
```
Comment (fasaxc) [inline felix/fv/rpf_test.go:39]: Think this needs `_BPF-SAFE_` in the name for it to be picked up by the BPF runs

File: felix/fv/rpf_test.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+				Expect(err).NotTo(HaveOccurred())
+
+				// Expect to see the packet from the .20 network at eth20 before RPF
+				Eventually(func() int { return tcpdumpHEP.MatchCount("UDP-30446") }, "1s", "100ms").
+					Should(BeNumerically("==", 1), "HEP - "+matcherHEP)
+
+				// Expect not to receive the packet from the .20 as it is dropped by RPF.
+				Consistently(func() int { return tcpdumpWl.MatchCount("UDP-30446") }, "1s", "100ms").
+					Should(BeNumerically("==", 0), "Wl - "+matcherWl)
+			})
```
Comment (fasaxc) [inline felix/fv/rpf_test.go:176]: Think you need a test where you set BPFRPF to disabled and verify that you get the opposite behaviour, otherwise, we could easily break the test so that the packets don't get through for some other reason.

File: felix/rules/static.go [hunk 13]
```diff
@@ -998,39 +998,94 @@ func (r *DefaultRuleRenderer) StaticRawTableChains(ipVersion uint8) []*Chain {
...
-			},
-			// Now ensure that the packet is not tracked.
-			Rule{
-				Action: NoTrackAction{},
-			},
+func (r *DefaultRuleRenderer) StaticBPFModeRawChains(ipVersion uint8, wgEncryptHost, enforceRPF bool) []*Chain {
+
+	var rawRules, rpfRules []Rule
+
+	rpfChainName := ChainNamePrefix + "RPF"
```
Comment (fasaxc) [inline felix/rules/static.go:1005]: Suggest making a constant alongside the other names in rules_def.go. 

Nit: I'd use lower case `rpf`, the convention was meant to be that upper case was used only for top-level chains (to match the iptables convention).  Doesn't really matter, but that was the intention.

**Review by fasaxc — CHANGES_REQUESTED** [review]
Looking good.  Just a few minor comments.
